content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,97 @@
+
+Solution: The best solution would be to work around the calls being made but if they must be done then it would be better to implement a Jenkins plugin which is able to gather the data needed.
+
+== Shared Libraries
+
+=== Just say no to overriding built-in pipeline Steps
+
+Wherever possible stay away from customized/overwritten pipeline steps. 
+This is the process where the standard pipeline APIs like `sh` or `timeout` are overwritten using shared libraries. 
+It is dangerous to do this because the pipeline APIs can change at any time causing custom code to break or give different results than expected. 
+These changes can also become difficult to troubleshoot when something is breaking or causing issues. 
+So even if custom code has not changed that does not mean that after an API update it will keep working the same. 
+Lastly, because of the ubiquitous use of these steps throughout pipelines, if something is coded incorrectly/inefficiently the results can be catastrophic to Jenkins.
+
+=== Avoid large global variable declaration files
+
+Having large variable declaration file will load every variable for every pipeline and depending on concurrency can end up using large amounts of memory for very little to no benefit. 
+It would be better to load variables which are relevant to the current execution.
+
+=== Avoid very large shared libraries
+
+This will load the same shared library per-job that is currently executing which can lead to large memory overhead. 
+This can also cause a slow down of every pipeline since it will need to checkout a very large file every time before starting.
+
+== Additional FAQs
+
+=== How do I deal with Concurrency in pipelines?
+
+Try not to share workspaces across multiple pipeline executions or multiple distinct pipelines. 
+This can lead to either unexpected file modification within each pipeline or each workspace will get renamed unexpectedly
+
+Ideally shared volumes/disks are mounted somewhere else and the files are copied from one place to the current workspace. 
+Then when the build is done it can be copied back
+
+Build in distinct containers which create needed resources from scratch(cloud-type agents work great for this). 
+This will ensure that the build process happens from scratch every time and is easily repeatable. 
+If that is not going to work then make sure that either concurrency is disabled on the pipeline or use the [lockable resources plugin](https://plugins.jenkins.io/lockable-resources) to lock up the workspace when running on it and no other builds can use while it is locked.
```

Comment:
I really hesitate to recommend lockable-resources in a best practices guide, since it seems to cause a ton of problems for people, especially as they scale.

Reply:
how does it cause issues? the ephemeral locks functionality released last week fixed the issues we had with it: https://github.com/jenkinsci/lockable-resources-plugin/releases

Reply:
If people know what they are doing, then I think lockable resources is OK. The issue comes when people just use it arbitrarily. Maybe I should add a caution here as well though?

Reply:
@timja mostly the issues come when people lock things and dont realize they are locking them